tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -14630,5 +14532,263 @@ SpirvEmitter::doUnaryExprOrTypeTraitExpr(const UnaryExprOrTypeTraitExpr *expr) {
+                                      inst->getSourceLocation());
+
+    forEachSpirvField(
+        recordType, dyn_cast<StructType>(type),
+        [this, inst, &result](size_t spirvFieldIndex, const QualType &fieldType,
+                              const StructType::FieldInfo &fieldInfo) {
+          auto *field = spvBuilder.createCompositeExtract(
+              fieldType, inst, {fieldInfo.fieldIndex},
+              inst->getSourceLocation());
+          field->setLayoutRule(inst->getLayoutRule());
+          auto decomposedField = decomposeToScalars(field);
+
+          // See how we can improve the performance by avoiding this copy.
+          result.insert(result.end(), decomposedField.begin(),
+                        decomposedField.end());
+          return true;
+        });
+    return result;
+  }
+
+  else {
+    resultType->dump();
+    llvm_unreachable("Trying to decompose a type that we cannot decompose");
+  }
+  return {};
+}
+
+SpirvInstruction *
+SpirvEmitter::GenerateFromScalars(QualType type,
+                                  std::vector<SpirvInstruction *> &scalars,
+                                  SpirvLayoutRule layoutRule) {
+  QualType elementType;
+  uint32_t elementCount = 0;
+  uint32_t numOfRows = 0;
+  uint32_t numOfCols = 0;
+
+  if (isScalarType(type)) {
+    // If the type if bool with a non-void layout rule, then it should be
+    // treated as a uint.
+    assert(layoutRule == SpirvLayoutRule::Void &&
+           "If the layout type is not void, then we should cast to an int when "
+           "type is a boolean.");
+    QualType sourceType = scalars[0]->getAstResultType();
+    if (sourceType->isBooleanType() &&
+        scalars[0]->getLayoutRule() != SpirvLayoutRule::Void) {
+      sourceType = astContext.UnsignedIntTy;
+    }
+
+    SpirvInstruction *result = castToType(scalars[0], sourceType, type,
+                                          scalars[0]->getSourceLocation());
+    scalars.erase(scalars.begin());
+    return result;
+  } else if (isVectorType(type, &elementType, &elementCount)) {
+    assert(elementCount <= scalars.size());
+    std::vector<SpirvInstruction *> elements;
+    for (uint32_t i = 0; i < elementCount; ++i) {
+      elements.push_back(castToType(scalars[i], scalars[i]->getAstResultType(),
+                                    elementType,
+                                    scalars[i]->getSourceLocation()));
+    }
+    SpirvInstruction *result = spvBuilder.createCompositeConstruct(
+        type, elements, scalars[0]->getSourceLocation());
+    result->setLayoutRule(layoutRule);
+    scalars.erase(scalars.begin(), scalars.begin() + elementCount);
+    return result;
+  } else if (isMxNMatrix(type, &elementType, &numOfRows, &numOfCols)) {
+    std::vector<SpirvInstruction *> rows;
+    SourceLocation loc = scalars[0]->getSourceLocation();
+    QualType rowType = astContext.getExtVectorType(elementType, numOfCols);
+    for (uint32_t i = 0; i < numOfRows; i++) {
```

Comment:
Since HLSL is row-major and SPIR-V is column-major, I would have expected this to be transposed when reconstructing, but it's difficult for me to tell for sure if that's correct or not without a test. Unfortunately I don't see an existing matrix flat conversion test that would have caught it if it did change. Can you add one?

Reply:
I added at test. Please double check that my understanding is correct. It is hard to follow exactly what is going on. Basicly, it seems like I row or col major matricies can be treated the same way in that part of the code.

Reply:
That seems correct to me, thanks for double checking and adding a test.